library/decorators.py

```diff
@@ -0,0 +1,26 @@
+import functools
+from collections.abc import Collection, Mapping, Hashable
+from frozendict import frozendict
+
+class Decorators:
+    
+    @staticmethod
+    def _deep_freeze(thing):
+        if thing is None or isinstance(thing, str):
+            return thing
+        elif isinstance(thing, Mapping):
+            vals = ((k, Decorators._deep_freeze(v)) for k, v in thing.items())
+            return frozendict(vals)
+        elif isinstance(thing, Collection):
+            return tuple(Decorators._deep_freeze(i) for i in thing)
+        elif not isinstance(thing, Hashable):
+            raise TypeError(f"unfreezable type: '{type(thing)}'")
+        else:
+            return thing
+
+    @staticmethod
+    def deep_freeze_args(func):
```

Comment:
This is a decorator for methods that have to be cached but have parameters that are not hashable (eg. `dict`, `list`). By placing it above a method declaration, it changes those into things that can be hashed.

Reply:
I just read up on this because I wasn't aware that parameters needed to be hashable if they are to be cached. That's really cool!

Reply:
You need to be able to quickly decide if the parameters sent are the same as a prior query. Hashing is a common approach to this

Reply:
Got it, just quickly read up on this.